### PR TITLE
Support/metamask

### LINKED ISSUE #463: Provide a way to construct `ethWalletCall` to allow MetaMask interact with CENNZnet extrinsics
As per the Wiki here https://wiki.cennz.net/#/dapp-development-guides/EVM/Guide?id=using-an-ethereum-keypair-with-cennznet

### PR DESCRIPTION
**Pull request by KarishmaBothara**
Closes https://github.com/cennznet/api.js/issues/463
Detail - with this PR, user can use CENNZnet API to sign tx with Metamask wallet. 
Users eth address should have CPAY to do a CENNZnet operation. The call required is as simple as
```
await api.tx.genericAsset
            .transfer(stakingAssetId, alice.address, transferAmt)
            . signViaEthWallet(
                ethAddress,
                api
            );
```
ethAddress - Ethereum address on metamask

### CODE REVIEW

File: packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts [hunk 1]
```diff
@@ -79,4 +82,28 @@ export default class CENNZnetExtrinsic extends GenericExtrinsic{
...
   }
+
+  private cvmToAddress(cvmAddress) {
+    let message = stringToU8a('cvm:');
+    message = u8aConcat(message, new Array(7).fill(0), hexToU8a(cvmAddress));
+    const checkSum = message.reduce((a, b) => a ^ b, 0);
+    message = u8aConcat(message, new Array(1).fill(checkSum));
+
+    return encodeAddress(message, 42);
+  }
```
Comment (jordy25519) [inline packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts]: should make this function available to package users

File: packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts [hunk 2]
```diff
@@ -79,4 +82,28 @@ export default class CENNZnetExtrinsic extends GenericExtrinsic{
...
+
+  /**
+   * @description sign the extrinsic via metamask
+   */
+  async signViaMetaMask(ethAddress: string, api: Api): Promise<SubmittableResultResult<ApiTypes>> {
```
Comment (jordy25519) [inline packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts]: name shouldn't be 'metamask'
other ethereum wallets can use this too: https://docs.walletconnect.com/json-rpc-api-methods/ethereum

**Review by jordy25519 — COMMENTED** [review]
cool! drive by review :)

File: packages/types/src/interfaces/augment-types.ts [hunk 3]
```diff
@@ -355,6 +356,7 @@ declare module '@polkadot/types/types/registry' {
     EthTransactionRequest: EthTransactionRequest;
     EthTransactionSignature: EthTransactionSignature;
     EthTransactionStatus: EthTransactionStatus;
+    ethWalletCall: ethWalletCall;
```
Comment (ken-futureverse) [inline packages/types/src/interfaces/augment-types.ts]: nitpick but should this be `EthWalletCall` with capital `E`?

File: packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts [hunk 4]
```diff
@@ -79,4 +81,19 @@ export default class CENNZnetExtrinsic extends GenericExtrinsic{
...
   }
+
+  /**
+   * @description sign the extrinsic via metamask
+   */
+  async signViaEthWallet(ethAddress: string, api: Api): Promise<SubmittableResultResult<ApiTypes>> {
+    const cennznetAddress = cvmToAddress(ethAddress);
+    const nonce = await api.rpc.system.accountNextIndex(cennznetAddress);
+    const payload = this.registry.createType('ethWalletCall', { call: this, nonce }).toHex();
+    const ethereum = typeof (global as any).ethereum !== "undefined" ? (global as any).ethereum : (window as any).ethereum;
```
Comment (ken-futureverse) [inline packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts]: I still think we should have the `ethereum` as parameter, would make it so much easier to mock test plus it doesn't care if `ethereum` is from `global` or `window`

**Review by ken-futureverse — CHANGES_REQUESTED** [review]
should we have test for the new extrinsic method `signViaEthWallet` and the `cvmToAddress` function?

**Comment by KarishmaBothara** [comment]
> should we have test for the new extrinsic method `signViaEthWallet` and the `cvmToAddress` function?

we have one for signViaEthWallet here - https://github.com/cennznet/api.js/blob/support/Metamask/packages/api/test/e2e/tx.e2e.ts#L130
can add one for cvmToAddress function

**Review by ken-futureverse — APPROVED** [review]

**Pull request merged**